globals/prefixes.go

```diff
@@ -320,13 +322,26 @@ var resourceTypeToPrefixes map[resource.Type][]string = func() map[resource.Type
 	return ret
 }()
 
-// RegisterPrefixSubtype is called from various packages to register which
-// prefixes belong to their subtypes. This lets the subtypes stay in different
+// RegisterPrefixToResourceInfo is called from various packages to register
+// which prefixes belong to them, what types those represent, and any
+// domain and subtype information. This lets the subtypes stay in different
 // packages (important for the reflection introspection we do) while not
 // creating import loops.
-func RegisterPrefixSubtype(prefix string, subtype Subtype) {
-	resInfo := prefixToResourceType[prefix]
+//
+// If called more than once for the same prefix, this function will panic.
+func RegisterPrefixToResourceInfo(prefix string, res resource.Type, domain string, subtype Subtype) {
+	resInfo, ok := prefixToResourceType[prefix]
+	if !ok && domain != "test" {
+		panic(fmt.Sprintf("prefix %q being registered to domain %q type %q subtype %q but did not already exist in map", prefix, domain, res.String(), subtype.String()))
+	}
+	if domain != "test" &&
+		(resInfo.Domain != "" ||
+			resInfo.Subtype != UnknownSubtype) {
```

Comment:
Given this, is it invalid to attempt to register with an empty domain and/or UnknownSubtype? So should the following result in a panic as well?:

```go
globals.RegisterPrefixToResourceInfo("foo", resource.Foo, "", globals.UnknownSubtype)
```

Reply:
Empty domain is totally fine, but agree that UnknownSubtype shouldn't be allowed.

Reply:
Actually let me clarify that. We don't _currently_ have any need for this, but it's not technically incorrect. You could be registering a type that isn't already registered in the map. But as we don't have a need for this now I'll make it panic.